Implement invalidate for non-offscreen mode

Spacetech opened the pull request:
This PR implements the `webContents.invalidate` method for non-offscreen mode. It will schedule a paint to the native window.

An example use case for this method: A few games on Steam use Electron and [Greenworks](https://github.com/greenheartgames/greenworks). Steam games have an in-game overlay that shows up over the game, which users can toggle via a keybind. This overlay has many issues in electron apps. 

The main issue is that since the overlay is hooked into the renderer, the overlay is frozen / corrupted when the renderer isn't painting or if it's only painting small dirty regions. We used to implement some shady methods to force repaints, such as having multiple 1x1 pixel canvases refreshing, or translateZ(0)ing some elements.. but these hacks stopped working properly in newer Chromium versions.  See [Greenworks#50](https://github.com/greenheartgames/greenworks/issues/50) for detailed discussion.
Calling this invalidate method often (60fps / every animation frame) completely solves this issue.

kevinsawicki commented:
@Spacetech what would be the best way to test this out to see it resolving the Greenworks issue you mentioned? I've never used Greenworks before and don't have any Steam games using Electron currently, thanks.

Spacetech commented:
Testing it is simple since Steam support non-steam games. I setup a repo with instructions on how to test it; https://github.com/Spacetech/electronsteamoverlaytest

Greenworks isn't related to the overlay issue itself. I brought it up since most Node.js based Steam games use it and because that Electron + Steam Overlay issue on their tracker was quite active.

Let me know if you have any questions. Thanks!

kevinsawicki commented:
Just a heads up, looks like this doesn't compile on macOS, `getNativeWindow` returns an `NSWindow` on that platform:

```
../../atom/browser/api/atom_api_web_contents.cc:1502:45: error: member access into incomplete type 'NSWindow'
      const gfx::Rect& bounds = nativeWindow->bounds();
                                            ^
/Users/travis/build/electron/electron/vendor/brightray/vendor/download/libchromiumcontent/src/ui/gfx/native_widget_types.h:88:7: note: forward declaration of 'NSWindow'
class NSWindow;
      ^
../../atom/browser/api/atom_api_web_contents.cc:1503:19: error: member access into incomplete type 'NSWindow'
      nativeWindow->SchedulePaintInRect(
                  ^
/Users/travis/build/electron/electron/vendor/brightray/vendor/download/libchromiumcontent/src/ui/gfx/native_widget_types.h:88:7: note: forward declaration of 'NSWindow'
class NSWindow;
```

kevinsawicki commented on atom/browser/api/atom_api_web_contents.cc:
```diff
@@ -1489,13 +1489,21 @@ int WebContents::GetFrameRate() const {
+                                            nullptr;
+    if (nativeWindow) {
+      const gfx::Rect& bounds = nativeWindow->bounds();
+      nativeWindow->SchedulePaintInRect(
```
Could this just use `SchedulePaint()` which paints the whole view, https://cs.chromium.org/chromium/src/ui/views/view.cc?l=831&rcl=54520ce5e56cd898668f0f86172cdeb9914a9b71

Spacetech replied:
I originally wanted to do that, but on windows `nativeWindow` is of type `aura::Window` and `SchedulePaint()` is a private member. https://cs.chromium.org/chromium/src/ui/aura/window.h?q=schedulepaint&dr=CSs&l=181

`c:\programming\electron\atom\browser\api\atom_api_web_contents.cc(1502): error C2248: 'aura::Window::SchedulePaint': cannot access private member declared in class 'aura::Window'
c:\programming\electron\vendor\brightray\vendor\download\libchromiumcontent\src\ui\aura\window.h(375): note: see declaration of 'aura::Window::SchedulePaint'
c:\programming\electron\vendor\brightray\vendor\download\libchromiumcontent\src\ui\aura\window.h(69): note: see declaration of 'aura::Window'`

kevinsawicki replied:
Ah, good point 👍

Spacetech commented:
Thanks for letting me know. I implemented a mac version that should work, however I can't test it since I don't have a mac. Also the Steam Overlay with Electron never worked on mac so I'm not sure how I would test it.

I added an `Invalidate` method to the `NativeWindow` class.
Looking at the apis, it seems like `invalidate` should be added to `BrowserWindow` instead of `WebContents`. But both having an invalidate method might be weird. Thoughts?

kevinsawicki commented:
>  But both having an invalidate method might be weird. Thoughts?

I think it is okay for now to have it only publicly available on the `webContents` even though it just delegates to the owner window in non-offscreen mode. Since this API is relatively new it feels like something that might still be tweaked as people start using it and we can revisit the best home for it in the future as it evolves.

kevinsawicki commented:
> Also the Steam Overlay with Electron never worked on mac so I'm not sure how I would test it.

I tried and couldn't get it working, I think it is okay to ship this in its current form though.

kevinsawicki commented:
> Testing it is simple since Steam supports non-steam games. I setup a repo with the instructions; https://github.com/Spacetech/electronsteamoverlaytest

Thanks for providing this, was able to build this branch on Windows 10 and verify the Steam overlay shows up and can be interacted with 👍

kevinsawicki commented:
Great job on this @Spacetech, thanks for adding support for it :ship: :shipit: 👍

Spacetech commented:
Thanks for reviewing and merging it

hokein commented:
@Spacetech Awesome! Just see it now (I should have noticed it earlier). I will check whether this solution works on OS X and Linux.

vaughnroyko commented:
@hokein Unfortunately the overlay will not work on OS X and Linux, even with this change, due to some changes that happened internally in Chromium at some point. I haven't found the version in which it broke in Electron; however, we found the version earlier in NW.js: https://github.com/greenheartgames/greenworks/issues/28 - Would love if we could track down the cause though!
